File: src/service/request-types.ts
/**
 * Service request types — shapes the client provides to the service.
 *
 * Separated from response types so the "what you send" contract is
 * distinct from the "what you get back" contract. When this grows
 * (auth, pagination, etc.) it can be promoted to a requests/ folder.
 */

import type { Seat, Vulnerability } from "../engine/types";
import type { OpponentMode, PlayPreference, PlayProfileId, PracticeMode, PracticeRole, SystemConfig, VulnerabilityDistribution } from "./session-types";

// ── Drill Handle ───────────────────────────────────────────────────

/** Opaque drill session identifier. */
export type DrillHandle = string;

// ── Target Selector ────────────────────────────────────────────────

/**
 * Drill target selector — picks which module / surface a drill is
 * focused on. `any` means no filter (any bundle-member surface is
 * eligible). `module` restricts to a single module. `surface` further
 * pins down a specific meaning id within that module.
 *
 * Discriminated union; mirrors Rust `TargetSelector` (`#[serde(tag =
 * "kind")]`).
 */
export type TargetSelector =
  | { kind: "any" }
  | { kind: "module"; moduleId: string }
  | { kind: "surface"; moduleId: string; surfaceId: string };

// ── Session Config ──────────────────────────────────────────────────

/**
 * Configuration for creating a new drill session.
 *
 * `systemConfig` and `baseModuleIds` are always provided by the caller.
 * Presets and custom systems use the same path — Rust never looks up configs by ID.
 * The TS layer resolves the selected system via `resolveSystemForSession()`.
 */
export interface SessionConfig {
  readonly conventionId: string;
  readonly userSeat?: Seat;
  readonly seed?: number;
  /** Full system configuration — always provided by the caller. */
  readonly systemConfig: SystemConfig;
  /** Base module IDs for this session. */
  readonly baseModuleIds: string[];
  /** Practice mode — controls auction entry point and play coupling. */
  readonly practiceMode?: PracticeMode;
  /** Drill target — selects which module / surface this drill is focused on. Absent === { kind: "any" }. */
  readonly target?: TargetSelector;
  /** Practice role — opener, responder, or both (random per deal). */
  readonly practiceRole?: PracticeRole;
  /** Play preference — whether to skip, prompt, or always play after bidding. */
  readonly playPreference?: PlayPreference;
  /** Opponent behavior mode. */
  readonly opponentMode?: OpponentMode;
  /** Vulnerability override. */
  readonly vulnerability?: Vulnerability;
  /** Opponent play skill — selects MC+DDS profile. */
  readonly playProfileId?: PlayProfileId;
  /** Probability weights over the four vulnerability states; the deal generator samples from this. */
  readonly vulnerabilityDistribution?: VulnerabilityDistribution;
}

File: crates/bridge-service/src/request_types.rs
//! Service request types — shapes the client provides to the service.

use bridge_conventions::types::system_config::SystemConfig;
use bridge_engine::types::{Seat, Vulnerability};
use serde::{Deserialize, Serialize};

use bridge_session::types::{OpponentMode, PlayPreference, PracticeMode, PracticeRole};

/// Opaque session identifier.
pub type DrillHandle = String;

/// Configuration for creating a new drill session.
///
/// `system_config` and `base_module_ids` are always provided by the caller.
/// Presets and custom systems use the same path — Rust never looks up configs by ID.
/// The TS layer resolves the selected system via `resolveSystemForSession()`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SessionConfig {
    pub convention_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user_seat: Option<Seat>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub seed: Option<u64>,
    /// Full system configuration — always provided by the caller.
    /// Presets and custom systems use the same path: the TS layer resolves
    /// the selected system to a full config before sending.
    pub system_config: SystemConfig,
    /// Base module IDs for this session. Presets use the standard 4;
    /// custom systems may differ.
    pub base_module_ids: Vec<String>,
    /// Practice mode — controls auction entry point and play coupling.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub practice_mode: Option<PracticeMode>,
    /// Target module for practice focus derivation.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub target_module_id: Option<String>,
    /// Practice role — opener, responder, or both (random per deal).
    #[serde(skip_serializing_if = "Option::is_none")]
    pub practice_role: Option<PracticeRole>,
    /// Play preference — whether to skip, prompt, or always play after bidding.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub play_preference: Option<PlayPreference>,
    /// Opponent behavior mode.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub opponent_mode: Option<OpponentMode>,
    /// Vulnerability override.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub vulnerability: Option<Vulnerability>,
}
